frontend/expr_inferrer.py

```diff
@@ -375,33 +376,74 @@ def _get_args_types(args, context, instance, solver):
 
 
-def _infer_annotated_function_call(args_types, solver, annotations):
-    result_type = solver.new_z3_const("call")
-    solver.annotation_resolver.add_annotated_function_axioms(args_types, solver, annotations, result_type)
+def _infer_annotated_function_call(args_types, solver, annotations, result_type):
+    return solver.annotation_resolver.get_annotated_function_axioms(args_types, solver, annotations, result_type)
 
-    return result_type
+
+def _infer_builtin_method(args_types, solver, context, result_type, method_name):
+    """Get the axioms of built-in methods inference"""
+    possible_methods = context.get_matching_methods(method_name)
+    method_axioms = []
+    for method in possible_methods:
+        cur_method_axioms = _infer_annotated_function_call(args_types, solver, method, result_type)
+        if cur_method_axioms is not None:
+            method_axioms.append(cur_method_axioms)
+    return method_axioms
 
 
 def infer_func_call(node, context, solver):
     """Infer the type of a function call, and unify the call types with the function parameters"""
     instance = None
+    result_type = solver.new_z3_const("call")
+    call_axioms = []
     if isinstance(node.func, ast.Attribute):
-        called, instance = infer(node.func, context, solver, True)
+        # Add axioms for built-in methods
+        instance = infer(node.func.value, context, solver)
         args_types = _get_args_types(node.args, context, instance, solver)
+        call_axioms += _infer_builtin_method(args_types, solver, context, result_type, node.func.attr)
     else:
         args_types = _get_args_types(node.args, context, instance, solver)
         if isinstance(node.func, ast.Name) and context.has_annotated_func(node.func.id):
-            return _infer_annotated_function_call(args_types, solver, context.get_annotated_func(node.func.id))
+            # Add axioms for built-in functions
+            annotated_func = context.get_annotated_func(node.func.id)
+            func_axioms = _infer_annotated_function_call(args_types, solver, annotated_func, result_type)
+            if func_axioms is not None:
+                call_axioms.append(func_axioms)
+
+    try:
+        # Add normal call inference axioms
         called = infer(node.func, context, solver)
-
-    result_type = solver.new_z3_const("call")
-
-    solver.add(axioms.call(called, args_types, result_type, solver.z3_types),
+        call_axioms += axioms.call(called, args_types, result_type, solver.z3_types)
+    except NameError as e:
+        # function name wasn't found in the context
+        if not call_axioms:
+            # no match for the function name in the context or in the built-ins
+            raise e
+
+    solver.add(Or(call_axioms),
                fail_message="Call in line {}".format(node.lineno))
 
     return result_type
 
 
+def _get_builtin_attr_access(instance_type, attr, result_type, context, solver):
+    """Return axioms for built-in attribute access"""
```

Comment:
Yes.

When we have the following attribute access for example:
```python
some_instance.some_method()
```

There are two possible scenarios:
- `some_instance` is a built in type, so we then should look for `some_method` in our stubs
- `some_instance` is a normal user defined instance.

At such point in the inference, we don't know which of each is the right one, so we try to follow both scenarios and `or` the resulting axioms. As a result, we attempt to infer `some_instance.some_method` as a normal attribute access with Z3. and then infer `some_instance.some_method()` as a normal function call. If we don't have this function, the first attribute access inference will add constrains on `some_instance` that it's only a user defined type, whereas it may be a built_in type.
So to sum it up, `_get_builtin_attr_access` is concerned with the attribute access inference, whereas `_infer_builtin_method` is concerned with the function call inference.